feat: New Users, Attendance Registration Improved and Optimized

nicolasantolini opened the pull request:
Enhance session management and data generation logic

- Updated `FeedController` to retrieve upcoming sessions for teachers, with improved logging.
- Introduced `SetAttendance` method in `SessionsController` for handling multiple attendance records, including validation and logging.
- Added `GetUpcomingSessionsForTeacher` method in `AttendanceService` and modified `SetTeacherApproval` to use student IDs.
- Refactored `MockDataGenerator` to use real user data instead of random generation, renaming the method to `GenerateRealData` and updating course creation logic.

nicolasantolini committed:
Enhance session management and data generation logic

- Updated `FeedController` to retrieve upcoming sessions for teachers, with improved logging.
- Introduced `SetAttendance` method in `SessionsController` for handling multiple attendance records, including validation and logging.
- Added `GetUpcomingSessionsForTeacher` method in `AttendanceService` and modified `SetTeacherApproval` to use student IDs.
- Refactored `MockDataGenerator` to use real user data instead of random generation, renaming the method to `GenerateRealData` and updating course creation logic.

nicolasantolini committed:
Enhance SetAttendance to support multiple records

Updated the SetAttendance method in SessionsController to accept a list of UpdateAttendanceContract, allowing for the processing of multiple attendance records. Adjusted validation to require at least one record and moved authorization checks before processing. Added logging for request processing.

joachimdekker commented on AttendanceSystem.Api/Controllers/SessionsController.cs:
```diff
@@ -142,10 +142,51 @@ public async Task<IActionResult> SetAttendance([HttpTrigger(AuthorizationLevel.A
-    
+
+    [Function($"{nameof(SessionsController)}-{nameof(SetAttendance)}")]
+    public async Task<ActionResult<Session>> SetAttendance([HttpTrigger(AuthorizationLevel.Anonymous, "put", Route = "courses/{courseId}/sessions/{sessionId:guid}/attendance")] HttpRequest req, [SwaggerIgnore] FunctionContext ctx, string courseId, Guid sessionId, [FromBody] List<UpdateAttendanceContract> contracts)
```
Did you try this out? I would expect swagger to reject generating the OpenAPI spec because you have a duplicate method.

Why is there a duplicate method? I already updated the `UpdateAttendanceContract` to have a list of contracts. I think you can merge this method with the one above, or at least only use one `UpdateAttendanceContract`.

joachimdekker commented on AttendanceSystem.Domain.Services/AttendanceService.cs:
```diff
@@ -40,6 +40,13 @@ public async Task<List<Session>> GetUpcomingSessionsForUser(string userId)
             .Where(s => s.Course.Students.Contains(userId)).ToListAsync();
     }
 
+    public async Task<List<Session>> GetUpcomingSessionsForTeacher(string userId)
+    {
+        return await _context.Sessions
+            .Include(s => s.Course)
+            .Where(s => s.Course.Teachers.Contains(userId)).OrderBy(m=>m.StartTime).ToListAsync();
+    }
```
I think we can refactor this method into the one above, getting:
```suggestion
    public async Task<List<Session>> GetUpcomingSessions(string userId)
    {
        return await _context.Sessions
            .Include(s => s.Course)
            .Where(s => s.Course.Students.Contain(userId) || s.Course.Teachers.Contains(userId)).OrderBy(m=>m.StartTime).ToListAsync();
    }
```

joachimdekker commented on AttendanceSystem.Domain.Services/AttendanceService.cs:
```diff
@@ -64,10 +71,10 @@ public async Task SetStudentAttendance(Guid sessionId, string studentId, Attenda
     {
         var session = await GetSessionWithRegister(sessionId);
-        session.SetTeacherApproval(teacherId, kind);
+        session.SetTeacherApproval(studentId, kind);
```
Excellent find!

joachimdekker commented on AttendanceSystem.Domain.Services/Tools/MockDataGenerator.cs:
```diff
@@ -87,76 +60,97 @@ public MockDataGenerator(UserService userService, CourseService courseService, A
     };
 
-        public async Task GenerateMockData()
+        public async Task GenerateRealData()
```
This is a great method! We may want to make this method idempotent in the future, but seeing as we don't have a lot of time, let's do that another time :)

joachimdekker commented on AttendanceSystem.Api/Controllers/FeedController.cs:
```diff
@@ -31,10 +32,22 @@ public async Task<IActionResult> GetUpcomingSessions([HttpTrigger(AuthorizationL
+            courses = await _attendanceService.GetUpcomingSessionsForTeacher(userId);
+            return new OkObjectResult(courses);
+        }
+        
```
Double empty-line?
```suggestion
```

joachimdekker reviewed (changes requested):
Excellent work. Left a few comments so we can simplify the code.

nicolasantolini committed:
Update AttendanceSystem.Api/Controllers/FeedController.cs

Co-authored-by: Joachim Dekker <jdekker3008@gmail.com>

nicolasantolini committed:
Update AttendanceSystem.Domain.Services/AttendanceService.cs

Co-authored-by: Joachim Dekker <jdekker3008@gmail.com>

nicolasantolini committed:
Enhance SetAttendance to process multiple records

Updated the SetAttendance method in SessionsController to handle multiple attendance records. The logic now validates user IDs for each record and processes them accordingly for both teachers and students. The return type has been changed to return an OkObjectResult with session details.

nicolasantolini committed:
Merge branch 'Sprint4-Nicola' of https://github.com/dcs-group-2/attendance-app-backend into Sprint4-Nicola
